presto-main/src/main/java/com/facebook/presto/operator/ContinuousWorkUtils.java

```diff
@@ -0,0 +1,322 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.google.common.collect.AbstractIterator;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.concurrent.Immutable;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Comparator.comparing;
+import static java.util.Objects.requireNonNull;
+
+public final class ContinuousWorkUtils
+{
+    private ContinuousWorkUtils() {}
+
+    static <T> Iterator<T> toIterator(ContinuousWork<T> work)
+    {
+        return new AbstractIterator<T>()
+        {
+            @Override
+            protected T computeNext()
+            {
+                while (true) {
+                    if (work.process()) {
+                        if (work.isFinished()) {
+                            return endOfData();
+                        }
+
+                        return work.getResult();
+                    }
+                    else if (work.isBlocked()) {
+                        try {
+                            work.getBlockedFuture().get();
+                        }
+                        catch (InterruptedException e) {
+                            Thread.interrupted();
+                            throw new RuntimeException(e);
+                        }
+                        catch (ExecutionException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                }
+            }
+        };
+    }
+
+    static <T> ContinuousWork<T> fromIterator(Iterator<T> iterator)
+    {
+        return create(() -> {
+            if (!iterator.hasNext()) {
+                return WorkState.finished();
+            }
+
+            return WorkState.ofResult(iterator.next());
+        });
+    }
+
+    static <T> ContinuousWork<T> mergeSorted(Iterable<ContinuousWork<T>> workIterable, Comparator<T> comparator)
```

Comment:
It is unexpected to see this method in this `commit`. It seems that it is part of `MergeHashSort` implementation.